pkg/apis/core/v1beta1/validation/validation.go

```diff
@@ -17,30 +17,117 @@ limitations under the License.
+	if spec.StatusType != nil {
+		allErrs = append(allErrs, ValidateStatusAPIResource(spec.StatusType, fldPath.Child("statusType"))...)
+	}
+
+	if spec.StatusCollection != nil {
+		allErrs = append(allErrs, validateEnumStrings(fldPath.Child("statusCollection"), string(*spec.StatusCollection), []string{string(v1beta1.StatusCollectionEnabled), string(v1beta1.StatusCollectionDisabled)})...)
+	}
+
+	return allErrs
+}
+
+const (
+	domainWithAtLeastOneDot string = "should be a domain with at least one dot"
+)
+
+func ValidateFederatedAPIResource(fedType *v1beta1.APIResource, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	allErrs = append(allErrs, ValidateAPIResource(&object.Spec.TargetType, field.NewPath("spec", "targetType"))...)
+	if len(fedType.Group) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("group"), ""))
+	} else if len(strings.Split(fedType.Group, ".")) < 2 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), fedType.Group, domainWithAtLeastOneDot))
+	}
 
 	return allErrs
 }
 
-func ValidateKubeFedCluster(object *v1beta1.KubeFedCluster) field.ErrorList {
+func ValidateStatusAPIResource(statusType *v1beta1.APIResource, fldPath *field.Path) field.ErrorList {
+	return ValidateFederatedAPIResource(statusType, fldPath)
+}
+
+func ValidateAPIResource(obj *v1beta1.APIResource, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+
+	if len(obj.Group) != 0 {
+		if errs := valutil.IsDNS1123Subdomain(obj.Group); len(errs) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), obj.Group, strings.Join(errs, ",")))
+		}
+	}
+
+	if len(obj.Version) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("version"), ""))
+	} else if errs := valutil.IsDNS1035Label(obj.Version); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("version"), obj.Version, strings.Join(errs, ",")))
+	}
+
+	if len(obj.Kind) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("kind"), ""))
+	} else if errs := valutil.IsDNS1035Label(strings.ToLower(obj.Kind)); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("kind"), obj.Kind, strings.Join(errs, ",")))
+	}
+
+	if len(obj.PluralName) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("pluralName"), ""))
+	} else if errs := valutil.IsDNS1035Label(obj.PluralName); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("pluralName"), obj.PluralName, strings.Join(errs, ",")))
+	}
+
+	allErrs = append(allErrs, validateEnumStrings(fldPath.Child("scope"), string(obj.Scope), []string{string(apiextv1b1.ClusterScoped), string(apiextv1b1.NamespaceScoped)})...)
+
 	return allErrs
 }
 
-func ValidateAPIResource(object *v1beta1.APIResource, fldPath *field.Path) field.ErrorList {
+func validateEnumStrings(fldPath *field.Path, value string, accepted []string) field.ErrorList {
```

Comment:
(No action required) The lack of reusable validation machinery seems problematic.

Reply:
No, not a comment on you (hence 'no action...').  It exists in apimachinery but it's not exported.